### PR TITLE
[ESIMD] Get rid of __SYCL_EXPLICIT_SIMD__ macro

### PR DESCRIPTION
**Pull request by DenisBakhvalov**

### CODE REVIEW

**Review by premanandrao — COMMENTED** [review]
The InitPreprocessor.cpp change looks good to me.   Someone else should review the header file changes.

Do we need a test to check the absence of the macro?

File: sycl/include/CL/sycl/detail/image_ocl_types.hpp [hunk 1]
```diff
@@ -207,6 +206,7 @@ template <access::mode AccessMode> struct opencl_image1d_buffer_type;
 __SYCL_IMAGE_BUFFER_TY_DEFINE(read, ro);
 __SYCL_IMAGE_BUFFER_TY_DEFINE(write, wo);
 __SYCL_IMAGE_BUFFER_TY_DEFINE(discard_write, wo);
+__SYCL_IMAGE_BUFFER_TY_DEFINE(discard_read_write, rw);
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/image_ocl_types.hpp:209]: Why this change is needed? The comment above says `__SYCL_IMAGE_BUFFER_TY_DEFINE` is not used anymore.

File: sycl/test/abi/layout_accessors_device.cpp [hunk 2]
```diff
@@ -29,6 +29,7 @@ SYCL_EXTERNAL void hostAcc(accessor<int, 1, access::mode::read, access::target::
 // CHECK-NEXT: 16 |         size_t [1] common_array
 // CHECK-NEXT: 24 |   union sycl::accessor<int, 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>
 // CHECK-NEXT: 24 |     sycl::accessor<int, 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>::ConcreteASPtrType MData
+// CHECK-NEXT: 24 |     sycl::accessor<int, 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>::OCLImage1dBufferTy ImageBuffer
```
Comment (kbobrovs) [inline sycl/test/abi/layout_accessors_device.cpp:32]: Please clarify if the ImageBuffer field was present before the change. Otherwise this would be ABI-breaking.

Comment (DenisBakhvalov) [reply sycl/test/abi/layout_accessors_device.cpp]: Ok. I reverted this change and instead of breaking the ABI of SYCL device code, my patch now breaks ABI of ESIMD device code. Hope it is OK for us to do.

File: sycl/include/CL/sycl/detail/image_ocl_types.hpp (hunk 1 above)
Comment (DenisBakhvalov) [reply sycl/include/CL/sycl/detail/image_ocl_types.hpp]: This change was needed a there was a specialization of accessor class with `discard_read_write`. Since we "unguard" this code from `-fsycl-explicit-simd` I need to add that.

Anyway, with my recent changes, this code is gone.

**Review by kbobrovs — APPROVED** [review]
LGTM

**Review by premanandrao — APPROVED** [review]
CFE changes LGTM

**Comment by DenisBakhvalov** [comment]
@bader, @kbobrovs, please merge.

**Comment by DenisBakhvalov** [comment]
@romanovvlad, can you please review on behalf of @intel/llvm-reviewers-runtime?

**Review by romanovvlad — APPROVED** [review]
